include/enumerations/material_definitions.hpp

```diff
@@ -0,0 +1,166 @@
+  constexpr std::array<
+      std::tuple<specfem::dimension::type, specfem::element::medium_tag,
+                 specfem::element::property_tag,
+                 specfem::element::boundary_tag>,
+      total_element_types>
+      material_systems{ MAKE_CONSTEXPR_ARRAY(ELEMENT_TYPES) };
+
+  return material_systems;
+}
+
+#undef MAKE_CONSTEXPR_ARRAY
+#undef MAKE_ARRAY_ELEM
+
+// Touch the following code at your own risk
+
+#define ELEM_IN_TUPLE(s, elem, tuple)                                          \
+  BOOST_PP_IF(                                                                 \
+      BOOST_PP_EQUAL(BOOST_PP_TUPLE_SIZE(tuple), 4),                           \
+      BOOST_PP_IF(                                                             \
+          BOOST_PP_EQUAL(GET_ID(s, _, BOOST_PP_TUPLE_ELEM(0, tuple)),          \
+                         GET_ID(s, _, BOOST_PP_TUPLE_ELEM(0, elem))),          \
+          BOOST_PP_IF(                                                         \
+              BOOST_PP_EQUAL(GET_ID(s, _, BOOST_PP_TUPLE_ELEM(1, tuple)),      \
+                             GET_ID(s, _, BOOST_PP_TUPLE_ELEM(1, elem))),      \
+              BOOST_PP_IF(                                                     \
+                  BOOST_PP_EQUAL(GET_ID(s, _, BOOST_PP_TUPLE_ELEM(2, tuple)),  \
+                                 GET_ID(s, _, BOOST_PP_TUPLE_ELEM(2, elem))),  \
+                  BOOST_PP_IF(BOOST_PP_EQUAL(                                  \
+                                  GET_ID(s, _, BOOST_PP_TUPLE_ELEM(3, tuple)), \
+                                  GET_ID(s, _, BOOST_PP_TUPLE_ELEM(3, elem))), \
+                              1, 0),                                           \
+                  0),                                                          \
+              0),                                                              \
+          0),                                                                  \
+      0)
+
+#define OP_OR(s, state, elem) BOOST_PP_OR(state, elem)
+
+#define ELEM_IN_SEQUENCE(elem)                                                 \
+  BOOST_PP_SEQ_FOLD_LEFT(                                                      \
+      OP_OR, 0, BOOST_PP_SEQ_TRANSFORM(ELEM_IN_TUPLE, elem, ELEMENT_TYPES))
+
+#define WHERE(...) (BOOST_PP_VARIADIC_TO_SEQ(__VA_ARGS__))
+
+#define CALL_FOR_ONE_SEQUENCE(s, MACRO, elem)                                  \
+  BOOST_PP_IF(ELEM_IN_SEQUENCE((BOOST_PP_SEQ_ENUM(elem))), MACRO, EMPTY_MACRO) \
+  (GET_TAG(s, _, BOOST_PP_TUPLE_ELEM(0, (BOOST_PP_SEQ_ENUM(elem)))),           \
+   GET_TAG(s, _, BOOST_PP_TUPLE_ELEM(1, (BOOST_PP_SEQ_ENUM(elem)))),           \
+   GET_TAG(s, _, BOOST_PP_TUPLE_ELEM(2, (BOOST_PP_SEQ_ENUM(elem)))),           \
+   GET_TAG(s, _, BOOST_PP_TUPLE_ELEM(3, (BOOST_PP_SEQ_ENUM(elem)))))
+
+#define EMPTY_MACRO(...)
+
+#define CREATE_SEQ(r, elem) (elem)
+
+#define CALL_MACRO_FOR_ALL_ELEMENT_TYPES(MACRO, seq)                           \
```

Comment:
I think we need more comments for the future. It would be good to explain what this
macros is doing using the rest of the macros. Otherwise, you have to go through 
the code line by line trying to figure it out.